### PR TITLE
(DI-92 / 137) Go libral

### PR DESCRIPTION
**Pull request by parisiale**
- Adding a C wrapper that exposes functions for retrieving types, providers, and resources in JSON format; all functions receive a pointer to a C-string pointer that will eventually point to the results;
  - adding a Go library that processes and exposes types, providers, and resources - its functions returns structured data after parsing and validating the JSON results from the underlying C wrapper;
 - adding Go tests and docs for the new library;
 - adding new CMake options: 1) LIBRAL_LIBRAY_STATIC, for building a static libral library; 2) AUGEAS_STATIC for using a static libaugeas (on linux)

### CODE REVIEW

**Comment by parisiale** [comment]
@lutter this is our preliminary work for having a simple Go interface.
We're proposing only one change to the implementation, namely to the JSON emitter, so that we can obtain a string out of JsonContainer without printing on stdout.

File: go_wrapper/libral.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
+package libral
+
+/*
+#cgo LDFLAGS: -fPIC -L${SRCDIR}/ -lstdc++ /usr/local/lib/libaugeas.dylib /usr/local/lib/libral.a  /usr/local/lib/libcpp-hocon.a /usr/local/lib/libyaml-cpp.a /usr/local/lib/leatherman_execution.a /usr/local/lib/leatherman_logging.a /usr/local/lib/leatherman_locale.a /usr/local/lib/leatherman_ruby.a /usr/local/lib/leatherman_dynamic_library.a /usr/local/lib/leatherman_util.a /usr/local/lib/leatherman_file_util.a /usr/local/lib/leatherman_json_container.a /usr/local/lib/libboost_system-mt.a /usr/local/lib/libboost_log-mt.a /usr/local/lib/libboost_log_setup-mt.a /usr/local/lib/libboost_thread-mt.a /usr/local/lib/libboost_date_time.a /usr/local/lib/libboost_filesystem-mt.a /usr/local/lib/libboost_chrono-mt.a /usr/local/lib/libboost_regex-mt.a /usr/local/lib/libboost_atomic-mt.a /usr/local/lib/libboost_program_options-mt.a /usr/local/lib/libboost_locale-mt.a /usr/local/lib/libiconv.a
```
Comment (johnmccabe) [inline go_wrapper/libral.go]: Is `/usr/local/lib` the default ld_library_path on all distros, if so we can probably get rid of the explicit paths, failing that replacing with `-L/usr/local/lib` and the `-lXXX` syntax.

Comment (johnmccabe) [inline go_wrapper/libral.go]: Should pull in https://github.com/parisiale/libral/pull/6

For building for other targets, Go doesn't support conditionals in the preprocessor, instead you can set build constraints in the filename etc (https://golang.org/pkg/go/build/#hdr-Build_Constraints)

**Comment by lutter** [comment]
A few questions about this:

* is the data conversion doing C++ -> JSON -> Go just done to get something going quickly or is that how this sort of thing is generally done in Go ? Ultimately, it will probably be better to eliminate the JSON step in the middle (maybe)
* I don't understand why the new STATIC flags are needed; AFAICT, you always get a libral.a anyway, and building with LIBRAL_STATIC should make sure you link against static libs (but I m sure I am missing something)
* I am not opposed to the emitter changes, but am wondering if basing the Go provider on the emitters is really the right path as they are closely tied to what ralsh outputs, and therefore changes in the JSON output from ralsh might break the Go bindings; in hindsight, they should have probably gone in the exe/ directory rather than be part of the library itself.
* We'll need an update to HACKING.md since there are now additional prereqs for building. Eventually, we'll probably split bindings out of the main repo, but until we have a sane distribution story, I think we'll need to bite the bullet and keep them in here (might be nice to have a CMake option to turn building Go bindings on/off for convenience)

**Comment by parisiale** [comment]
Thanks!

> is the data conversion doing C++ -> JSON -> Go just done to get something going quickly or is that how this sort of thing is generally done in Go ?

The choice of using C-string & JSON serialization was made to cope with cgo's need for a C interface and the way it interoperates with C via pointers. If possible, we could instantiate a the new Go types directly in C++.
Can that be done @johnmccabe ?

> I don't understand why the new STATIC flags are needed; AFAICT, you always get a libral.a anyway, and building with LIBRAL_STATIC should make sure you link against static libs (but I m sure I am missing something)

The reason behind the new flags is that LIBRAL_STATIC doesn't work on OS X, where we started working. Also, I tried to have the same approach as for [facter](https://github.com/parisiale/facter/tree/go_wip).
Mind that we don't have OS X in our target platform list, so we can get rid of all that. @davejohnston, can we reset the CMake files and get it building on your linux box?

> I am not opposed to the emitter changes, but am wondering if basing the Go provider on the emitters is really the right path as they are closely tied to what ralsh outputs, and therefore changes in the JSON output from ralsh might break the Go bindings; in hindsight, they should have probably gone in the exe/ directory rather than be part of the library itself.

I'll have a look into that. Things would become simpler if we abandon the JSON serialization, ofc.
But, for the record, I'm for letting the Go wrapper propagate what the JSON emitter generates, without any validation / processing; the go-libral consumer would then process it as it pleases. In that way nothing would be coupled.

> We'll need an update to HACKING.md since there are now additional prereqs for building. Eventually, we'll probably split bindings out of the main repo, but until we have a sane distribution story, I think we'll need to bite the bullet and keep them in here (might be nice to have a CMake option to turn building Go bindings on/off for convenience)

Ok. I'll update once we decide whether we need new options and get one specific for building the Go bindings.

**Comment by johnmccabe** [comment]
@lutter / @parisiale re `C++ -> JSON -> Go` this was done for speed, so that we can up an running as quickly as possible.

Yes, its possible to map more directly to native types, and I think this is desirable, but my gut feeling would be to address this at a later point if you don't see it as a blocker for an initial merge.

**Comment by davejohnston** [comment]
I can confirm building without the changes to CMakesLists.txt  and just passing the LIBRAL_STATIC 
 option produces a static lib, that we can link to the go binary.

**Comment by johnmccabe** [comment]
@lutter are the thread-safe boost libs required? I can't see them in the docker based build?

**Comment by johnmccabe** [comment]
Are unit tests failing for any of you? I'm seeing this when building.
```
Running tests...
Test project /usr/src/libral/build
    Start 1: unit_tests
1/4 Test #1: unit_tests .......................***Failed    2.80 sec
    Start 2: smoke_test
2/4 Test #2: smoke_test .......................   Passed    2.47 sec
    Start 3: help_test
3/4 Test #3: help_test ........................   Passed    0.02 sec
    Start 4: version_test
4/4 Test #4: version_test .....................   Passed    0.02 sec

75% tests passed, 1 tests failed out of 4

Total Test time (real) =   5.37 sec

The following tests FAILED:
	  1 - unit_tests (Failed)
Errors while running CTest
make: *** [test] Error 8
```

**Comment by lutter** [comment]
@johnmccabe I am not explicitly pulling thread-safe boost libs in; I suspect that something in leatherman pulls them in/needs them

**Comment by johnmccabe** [comment]
@lutter I don't think they're needed, ralsh for example links against the non-mt (seems like -mt is a windows naming convention and there isn't a difference on Linux anyway). Not sure why they're in the LDFLAGS in the wrapper..

**Comment by johnmccabe** [comment]
I've some changes to push into this later tonigh, just a heads up

**Comment by parisiale** [comment]
I'm passing the ownership of this PR to @davejohnston ; I'll be ready to jump in as needed

File: HACKING.md [hunk 2]
```diff
@@ -171,3 +171,22 @@ Again, following the same instructions as Linux.
...
 ```
+
+## Building Go Wrapper
+
+The Go Wrapper exposes libral functions using a 'C' interface.  Currently
+ these methods are being used by golang code to link to libral.
+By default the cwrapper is not included in the libral library.  If you 
+want to enable it use the BUILD_GO option.  N.B this option can also be
+combined with other build options such as LIBRAL_STATIC to produce a static
+library.
```
Comment (lutter) [inline HACKING.md]: What do I need to have installed to build the go wrappers ? Just the compiler ?

Comment (johnmccabe) [inline HACKING.md]: Just docker, PR incoming with the updates

**Pull request merged**